Python/ceval.c

```diff
@@ -3786,13 +3787,21 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
                   We'll be passing `oparg + 1` to call_function, to
                   make it accept the `self` as a first argument.
                 */
+
+                PyObject **pfunc = (*(&sp)) - (oparg + 1) - 1;
+                PyObject *func = *pfunc;
+                PyCodeObject *co = (PyCodeObject *)PyFunction_GET_CODE(func);
                 res = call_function(tstate, &bounds, &sp, oparg + 1, NULL);
+                if (res == NULL && PyErr_ExceptionMatches(PyExc_TypeError)) {
```

Comment:
Can TypeError be raised for other reason?

Reply:
Yes, but the checks in `improve_missing_self_error` will trigger only on the case we want (the number of arguments to the function mismatch).

Reply:
I am moving the guard `nargs + 1 != co->co_argcount` here so is more clear when this triggers and to avoid the extra function call

Reply:
Done in deb282d5c

Reply:
If the number of arguments to the function mismatches *and* the trace function raises a TypeError, your "Did you ..." can be attached to the unrelated error message.